feat(KNO-13529): Docs for Knock hosted preference center (w/o custom domains) (#1474)
knocklabs/docs: content/preferences/hosted-preference-center.mdx

---
title: "Hosted preference center"
description: "Learn how to use Knock's hosted preference center to let your users manage notification preferences without writing any code."
tags:
  ["preferences", "preference center", "hosted", "unsubscribe", "commercial"]
section: Preferences
---

> Knock's hosted preference center is a page where your users manage their notification preferences. You configure it from the dashboard; Knock hosts and renders the page, so you can launch a preference center without building or hosting any UI yourself.

Reviewer: ```suggestion
Knock's hosted preference center is a page where your users manage their `default` notification preferences. You configure it from the dashboard; Knock hosts and renders the page, so you can launch a preference center without building or hosting any UI yourself.
```
I think we should add this to make clear that we don't support tenant-specific preferences. Alternatively we could make an update to the **Users only** callout to list caveats (i.e. you can't host a preference center for a tenant or an object, and it's only the user's `default`)

      <a href="/concepts/objects">objects</a> or{" "}
      <a href="/concepts/tenants">tenants</a>.
    </>
  }
/>

## How it works

Knock hosts the preference center at `p.knock.app`. Each user accesses their own preferences through a signed link that looks like this:

```
https://p.knock.app/?t=<token>
```

The token identifies the recipient, so users don't need to log in to your product to view or update their preferences.

When a user opens the link, they see their preference settings based on the configuration in your dashboard, with your account branding, preference options, and the ability to opt out of commercial messaging.

When the user clicks **Save preferences**, Knock writes their changes back to their `default` preference set. The link never expires.

<Callout
  type="info"
  title="Conditional preferences."
  text={
    <>
      If a preference row, or a channel type within a row, currently has{" "}
      <a href="/preferences/preference-conditions">conditional</a> preferences
      for a given user, the preference center renders that key but the user
      can't modify their preference for it.
    </>
  }
/>

## Configure your preference center

> You configure the preference center from the Knock dashboard under **Platform** > **Preferences** > **Preference center**. Your configuration is environment-specific.

Reviewer: ```suggestion
You configure the preference center from the Knock dashboard under <a href="https://dashboard.knock.app/~/preferences?tab=preference_center" target="_blank">**Platform** > **Preferences** > **Preference center**</a>. Your configuration is environment-specific.
```

---
title: "Hosted preference center"
description: "Learn how to use Knock's hosted preference center to let your users manage notification preferences without writing any code."
tags:
  ["preferences", "preference center", "hosted", "unsubscribe", "commercial"]
section: Preferences
---

> Knock's hosted preference center is a page where your users manage their notification preferences. Knock hosts and renders the page, you configure it from the dashboard, so you can launch a preference center without building or hosting any UI yourself.

Reviewer: ```suggestion
Knock's hosted preference center is a page where your users manage their notification preferences. You configure it in your dashboard and Knock hosts and renders the page, so you can launch a preference center without building or hosting any UI yourself.
```

> Knock's hosted preference center is a page where your users manage their notification preferences. Knock hosts and renders the page, you configure it from the dashboard, so you can launch a preference center without building or hosting any UI yourself.

Reviewer: ```suggestion
Knock's hosted preference center is a page where your users manage their `default` notification preferences. Knock hosts and renders the page, you configure it from the dashboard, so you can launch a preference center without building or hosting any UI yourself.
```

> This is the fastest way to give your users a preference center. If you need more control over the experience, you can still [build your own preference center UI](/preferences/overview#build-your-preference-center) on top of Knock's preferences API.

Reviewer: ```suggestion
This is the fastest way to give your users access to updating their preferences when they receive a notification. If you need more control over the experience, you can still [build your own preference center UI](/preferences/overview#build-your-preference-center) on top of Knock's preferences API.
```

<Callout
  type="info"
  title="Users only."
  text="The hosted preference center currently supports user recipients only. It doesn't support managing preferences for objects or tenants."
/>

## How it works

Knock hosts the preference center at `p.knock.app`. Each user accesses their own preferences through a signed link that looks like this:

```
https://p.knock.app/?t=<token>
```

> The token identifies the recipient, so users don't need to log in to your product to view or update their preferences. When a user opens the link, they see their preference settings based on the environment-level preference center configuration shape. When the user clicks **Save preferences**, Knock writes their changes back to the recipient's `default` preference set. The link never expires.

Reviewer: ```suggestion
The token identifies the recipient, so users don't need to log in to your product to view or update their preferences. 

When a user opens the link, they see their preference settings based on the configuration in your dashboard with your account branding, preference options, and the ability to opt out of commercial messaging.

When the user clicks **Save preferences**, Knock writes their changes back to their `default` preference set. The link never expires.
```

> The page reflects the recipient it was generated for, including the account branding, the preference options you've configured, and a row for commercial messages opt-out.

Reviewer: Moved this into the previous paragraph
```suggestion
```

<Callout
  type="info"
  title="Preferences managed by conditions."
  text={
    <>
      If a preference row, or a channel type within a row, currently has{" "}

>       <a href="/preferences/preference-conditions">condition-managed</a>{" "}

Reviewer: ```suggestion
      <a href="/preferences/preference-conditions">conditional</a>{" "}
```

      preferences for a given user, the preference center renders that key but
      the user can't modify their preference for it.
    </>
  }
/>

> ## Enable and configure your preference center

Reviewer: Shortening for brevity and to fit better in the sidebar. I think "enabling" can be considered part of configuring and we'll have it as its own step below.
```suggestion
## Configure your preference center
```

You configure the preference center from the Knock dashboard under **Platform** > **Preferences** > **Preference center**. Your configuration is environment-specific.

> When the preference center is saved and enabled, changes go live in the environment right away. When the preference center is disabled, users can only edit their commercial messages opt-out preference.

Reviewer: ```suggestion
When the preference center is saved and enabled, changes go live in the environment right away. When the preference center is disabled, users can only edit their [commercial messages opt-out](/preferences/commercial-unsubscribe) preference.
```

<Steps titleSize="h3">

>   <Step title="Enable the preference center">

Reviewer: I think "Enable the preference center" should be the last step, even though it's at the top of the configuration screen.

  </Step>
  <Step title="Configure commercial messages">
    The commercial messages row gives your users a single place to opt out of promotional or non-essential notifications. This is backed by the same `commercial_subscribed` preference used for [commercial unsubscribe](/preferences/commercial-unsubscribe), so opting out here also satisfies one-click CAN-SPAM requirements for your commercial messaging.

    The commercial messages row is mandatory and always renders in the preference center, whether the preference center is enabled or disabled. You can change the title and body copy for the row, but you can't remove it.

  </Step>
  <Step title="Set your branding">
    The preference center uses the branding configured for your account, including your logo, icon, and primary color. See [branding](/template-editor/branding) for more on how branding works in Knock.
  </Step>
  <Step title="Preview as a user">

>     Before you save, you can preview the preference center as a specific user. This loads that user's literal preference center, rendering the page exactly as they would see it with their current preferences and your latest configuration, so you can confirm the experience before it goes live.

Reviewer: ```suggestion
    Before you save, you can preview the preference center as a specific user. This renders that user's page exactly as they would see it with their current preferences and your latest configuration, so you can confirm the experience before it goes live.
```

  text={
    <>
      The hosted preference center doesn't support top-level{" "}
      <a href="/preferences/overview#channels-vs-channel-types">channel type</a>{" "}
      preferences. Channel types can only be configured as part of a category or
      workflow row.
    </>
  }
/>

When a user saves their preferences, Knock updates the corresponding keys on the recipient's `default` preference set. These changes are merged with any environment or tenant defaults during [preference evaluation](/preferences/overview#preference-evaluation-rules), with the recipient's preferences taking precedence according to the [merge hierarchy](/preferences/overview#merging-preferences).

<Callout
  type="info"
  title="Multiple rows can affect one notification."
  text={
    <>
      A single notification can be affected by preferences set across several
      preference center rows. If a user sets any of those preferences to{" "}
      <code>false</code>, the notification won't be delivered. See{" "}
      <a href="/preferences/overview#preference-evaluation-rules">
        preference evaluation rules
      </a>{" "}
      for how Knock resolves conflicts.
    </>
  }
/>

> For example, imagine the `new-comment` workflow belongs to the `collaboration` category, and your preference center has a row for the `collaboration` category and a separate row for the `new-comment` workflow. If a user opts out of either the `collaboration` category row or the `new-comment` workflow row, the comment notification won't be delivered, because both preferences must evaluate to `true` for the notification to send.

Reviewer: ```suggestion
```
I'm open to differing opinions on this, but I think the link in the previous paragraph that points to how preference evaluation/merge hierarchy works is good enough, and that we don't need to re-document that on this page.

> ## Linking users to the preference center

Reviewer: nit: for brevity
```suggestion
## Linking to the preference center
```

> To send users to their preference center, include the manage-preferences URL in your email templates using the built-in variable:

Reviewer: ```suggestion
To link a user directly to their preference center from a notification, include the URL in your message template or [email layout](/integrations/email/layouts) using the built-in variable:
```

> ```liquid title="Link to the hosted preference center"

Reviewer: ```suggestion
```liquid title="Link to the hosted preference center in an email template"
```

<a href="{{vars.manage_preferences_url}}">Manage your preferences</a>
```

> This sits alongside the existing `vars.commercial_unsubscribe_url` variable, which links to the one-click [commercial unsubscribe](/preferences/commercial-unsubscribe) flow. Use `vars.manage_preferences_url` when you want to give users full control over their preferences, and `vars.commercial_unsubscribe_url` for a single-click opt-out of commercial messages.

Reviewer: ```suggestion
Knock will render the user-specific link in the recipient's notification on your behalf.

Use the `vars.manage_preferences_url` variable when you want to give users full control over their preferences via the hosted preference center. For single-click opt-out of commercial messages via the [commercial unsubscribe](/preferences/commercial-unsubscribe) feature, use `vars.commercial_unsubscribe_url` instead.
```

> You can also add the preference center link to your emails without editing template code by configuring footer links in your [email layout settings](/integrations/email/layouts).

Reviewer: Suggest cutting this separate mention in favor of briefer mention above.
```suggestion
```

## Copying configuration between environments

> Your preference center configuration is environment-specific, just like your [preference defaults](/preferences/overview#create-a-default-preferenceset). You can copy the configuration from one environment to another to keep your environments in sync as you move from development to production.

Reviewer: ```suggestion
Your preference center configuration is environment-specific, just like your [preference defaults](/preferences/overview#create-a-default-preferenceset). You can copy the configuration from one environment to another using the "Copy to..." button in the upper right corner of your preference center configuration. This enables you to keep your environments in sync as you move from development to production.
```

## Learn more

> To learn more about the preferences model behind the preference center, see the [preferences overview](/preferences/overview). To set up one-click unsubscribe for commercial messaging, see [commercial unsubscribe](/preferences/commercial-unsubscribe).

Reviewer: ```suggestion
To learn more about the preferences model behind the preference center, see the [preferences overview](/preferences/overview).
```
I think we can probably cut the one-click unsub mention here, given how many times we mention it further up the page.
